src/redis_lib/mod.rs

```diff
@@ -10,44 +11,134 @@ use tracing::error;
 ///
 /// * REDIS_PORT: the port of redis, defaults to 6379
 /// * REDIS_HOST: the host of redis, ip or hostname or domain, defaults to 127.0.0.1
-pub fn get_connection() -> Option<Client> {
-    let redis_port: u16;
-    match env::var("REDIS_PORT") {
-        Ok(val) => match val.parse() {
-            Ok(value) => redis_port = value,
-            Err(err) => {
-                error!(error=err.to_string(), "Failed to parse int");
-                return None;
-            }
-        },
-        Err(_) => redis_port = 6379
-    };
-    let redis_hostname = match env::var("REDIS_HOST") {
-        Ok(val) => val,
-        Err(_) => "127.0.0.1".to_string()
-    };
-    let client;
-    match Client::open(format!("redis://{}:{}", redis_hostname, redis_port)) {
-        Ok(val) => client = val,
+pub async fn get_connection() -> Option<(Client, Connection)> {
+    let redis_port: Result<u16, _> = env::var("REDIS_PORT").unwrap_or("6379".to_owned()).parse();
+    if let Err(err) = redis_port {
+        error!(error = err.to_string(), "Failed to parse int");
+        return None;
+    }
+    let redis_port = redis_port.unwrap();
+
+    let redis_hostname = env::var("REDIS_HOST").unwrap_or("127.0.0.1".to_owned());
+
+    let client = Client::open(format!("redis://{redis_hostname}:{redis_port}"))
+        .map_err(|e| {
+            error!(error = e.to_string(), "Problem creating redis client");
+        })
+        .ok()?;
+
+    let connection = match client.get_async_connection().await {
         Err(err) => {
-            error!(error=err.to_string(), "Problem creating redis client");
+            error!(error = err.to_string(), "Failed connecting to Redis");
             return None;
         }
+        Ok(con) => con,
+    };
+
+    Some((client, connection))
+}
+
+/// Increments the counter that tracks how often an email was successfully used to verify an account
+#[allow(unused)]
+pub async fn increment_email_used_verification_success(con: &mut Connection, email: &str) {
+    let email_hash = hash_email(email);
+    let current = get_email_stats(con, &email_hash).await;
+    set_email_stats(con, &email_hash, current.0 + 1, current.1);
+}
+
+/// Increments the counter that tracks how often an email was given to verify an account
+#[allow(unused)]
+pub async fn increment_email_used_verification(con: &mut Connection, email: &str) {
+    let email_hash = hash_email(email);
+    let current = get_email_stats(con, &email_hash).await;
+    set_email_stats(con, &email_hash, current.0, current.1 + 1);
+}
+
+/// gets how often an email was used to verify an account (not necessarily successful)
+#[allow(unused)]
+pub async fn get_email_used_verification_success(con: &mut Connection, email: &str) -> u64 {
+    get_email_stats(con, &hash_email(email)).await.0
+}
+
+/// gets how often an email was used to verify an account (not necessarily successful)
+#[allow(unused)]
+pub async fn get_email_used_verification(con: &mut Connection, email: &str) -> u64 {
```

Comment:
Document that a response of `0` may either be the value in redis or that the email was not saved in redis.